pkg/snmp/traps/formatter.go

```diff
@@ -0,0 +1,249 @@
+	}
+
+	data := make(map[string]interface{})
+
+	uptime, err := parseSysUpTime(variables[0])
+	if err != nil {
+		return nil, err
+	}
+	data["uptime"] = uptime
+
+	trapOID, err := parseSnmpTrapOID(variables[1])
+	if err != nil {
+		return nil, err
+	}
+	data["snmpTrapOID"] = trapOID
+
+	trapMetadata, err := f.oidResolver.GetTrapMetadata(trapOID)
+	if err != nil {
+		log.Debugf("unable to resolve OID: %s", err)
+	} else {
+		data["snmpTrapName"] = trapMetadata.Name
+		data["snmpTrapMIB"] = trapMetadata.MIBName
+	}
+
+	parsedVariables := parseVariables(variables[2:])
+	data["variables_raw"] = parsedVariables
+	for _, variable := range parsedVariables {
+		varMetadata, err := f.oidResolver.GetVariableMetadata(trapOID, variable.OID)
+		if err != nil {
+			log.Debugf("unable to enrich variable: %s", err)
+			continue
+		}
+		data[varMetadata.Name] = variable.Value
+	}
+	return data, nil
+}
+
+// NormalizeOID convert an OID from the absolute form ".1.2.3..." to a relative form "1.2.3..."
+func NormalizeOID(value string) string {
+	// OIDs can be formatted as ".1.2.3..." ("absolute form") or "1.2.3..." ("relative form").
+	// Convert everything to relative form, like we do in the Python check.
+	return strings.TrimLeft(value, ".")
+}
+
+func parseSysUpTime(variable gosnmp.SnmpPDU) (uint32, error) {
+	name := NormalizeOID(variable.Name)
+	if name != sysUpTimeInstanceOID {
+		return 0, fmt.Errorf("expected OID %s, got %s", sysUpTimeInstanceOID, name)
+	}
+
+	value, ok := variable.Value.(uint32)
+	if !ok {
+		return 0, fmt.Errorf("expected uptime to be uint32 (got %v of type %T)", variable.Value, variable.Value)
+	}
+
+	return value, nil
+}
+
+func parseSnmpTrapOID(variable gosnmp.SnmpPDU) (string, error) {
+	name := NormalizeOID(variable.Name)
+	if name != snmpTrapOID {
+		return "", fmt.Errorf("expected OID %s, got %s", snmpTrapOID, name)
+	}
+
+	value := ""
+	switch variable.Value.(type) {
+	case string:
+		value = variable.Value.(string)
+	case []byte:
+		value = string(variable.Value.([]byte))
+	default:
+		return "", fmt.Errorf("expected snmpTrapOID to be a string (got %v of type %T)", variable.Value, variable.Value)
+	}
+
+	return NormalizeOID(value), nil
+}
+
+func parseVariables(variables []gosnmp.SnmpPDU) []trapVariable {
+	var parsedVariables []trapVariable
+
+	for _, variable := range variables {
+		varOID := NormalizeOID(variable.Name)
+		varType := formatType(variable)
+		varValue := formatValue(variable)
+		parsedVariables = append(parsedVariables, trapVariable{OID: varOID, VarType: varType, Value: varValue})
+	}
+
+	return parsedVariables
+}
+
+func formatType(variable gosnmp.SnmpPDU) string {
+	switch variable.Type {
+	case gosnmp.Integer, gosnmp.Uinteger32:
+		return "integer"
+	case gosnmp.OctetString:
+		return "string"
+	case gosnmp.ObjectIdentifier:
+		return "oid"
+	case gosnmp.Counter32:
+		return "counter32"
+	case gosnmp.Counter64:
+		return "counter64"
+	case gosnmp.Gauge32:
+		return "gauge32"
+	default:
+		return "other"
+	}
```

Comment:
Do we need to expose the raw snmp types to user ?

Somehow the value should already contain the type e.g. string value (string type), int value (int type), etc

(no blocking for this PR since it was already there)

Reply:
Yeah I tried to keep how it was before and I don't see a real reason to remove it now that it's there

Reply:
Added a JIRA card for this and the snmp version